source/mir/random/tinflex/internal/calc.d

```diff
@@ -0,0 +1,154 @@
+module mir.random.tinflex.internal.calc;
+
+import mir.random.tinflex.internal.types : Interval;
+
+/**
+Calculate the mean between two points using the arcmean:
+
+    tan(0.5 * (atan(l) + atan(r))
+
+In contrast to the normal mean (`0.5 * (l + r)`) being a geometric plane,
+the arcmean favors the mean region more.
+
+Params:
+    l = Left point
+    r = Right point
+
+Returns:
+    Splitting point within the interval
+
+See_Also:
+    $(LINK2 http://www.wolframalpha.com/input/?i=tan(0.5+*+(ArcTan%5Bx%5D+%2B+ArcTan%5By%5D)),
+    WolframAlpha visualization of the arc-mean)
+
+References:
+    Hormann, W., J. Leydold, and G. Derflinger.
+    "Automatic Nonuniform Random Number Generation." (2004): Formula 4.23
+*/
+auto arcmean(S, bool sorted = false)(const S l, const S r)
+out (result)
+{
+    assert(iv.lx <= result && result <= iv.rx);
+}
+body
+{
+    import std.math: atan, tan;
+
+    S _l = l;
+    S _r = r;
+    static if (!sorted)
+    {
+        if (_r < _l)
+        {
+            S t = _r;
+            _r = _l;
+            _l = t;
+        }
+    }
+
+    if (_r < -S(1e3) || _l > S(1e3))
+        return 2 / (1 / _l + 1 / r);
+
+    immutable d = atan(_l);
+    immutable b = atan(_r);
+
+    assert(d <= b);
+    if (b - d < S(1e-6))
+        return S(0.5) * _l + S(0.5) * r;
+
+    return tan(S(0.5) * (d + b));
+}
+
+/**
+Calculate the exponential mean between an Interval `iv`.
+
+ \int_l^r x * (h(x) - s(x)) dx
+ -----------------------------
+ \int^l^r      h(x) - s(x)  dx
+
+It can't be applied for unbounded intervals if c <= -1/2
+
+References:
+    Hormann, W., J. Leydold, and G. Derflinger.
+    "Automatic Nonuniform Random Number Generation." (2004): Formula 4.22
+*/
+auto expmean(S)(in ref Interval!S iv)
+out (result)
```

Comment:
while looking for the correct reference of "arc-mean", I found the exp-mean (exp_expected new construction point) method.

![image](https://cloud.githubusercontent.com/assets/4370550/16508045/3c948fee-3f2f-11e6-8ef4-d99833286152.png)

I went ahead and implemented it in a very naive way to test its implications (please no performance critic yet).

However for the NormalDistribution with arcmean 34 intervals are required, with exp-mean 46 :/

`h(x)` and  `s(x)` are defined as linear functions with three parameters `f(x) = a * b * (x - y)`.
I checked my result with WolframAlpha:

upper: http://www.wolframalpha.com/input/?i=Integrate%5B(a_1%2Bb_1(x-y_1)+-+a_2%2Bb_2(x-y_2)),+x%5D

lower: http://www.wolframalpha.com/input/?i=Integrate%5Bx*+(a_1%2Bb_1(x-y_1)+-+a_2%2Bb_2(x-y_2)),+x%5D

So far I couldn't find more references of this method except this tiny excerpt from above, so I hope I interpreted it right.